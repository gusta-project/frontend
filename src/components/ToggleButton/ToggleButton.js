import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export const toggleButtonProps = {
  variants: ['check']
};

export default class ToggleButton extends Component {
  static propTypes = {
    className: PropTypes.string,
    buttonProps: PropTypes.object,
    iconProps: PropTypes.object,
    initialValue: PropTypes.bool,
    variant: PropTypes.oneOf(toggleButtonProps.variants),
    onClick: PropTypes.func
  };

  static defaultProps = {
    initialValue: false,
    variant: 'check'
  };

  constructor(props) {
    super(props);

    this.state = { value: this.props.initialValue };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(event) {
    const { onClick } = this.props;

    this.setState({ value: !this.state.value });

    if (onClick) {
      onClick(event);
    }
  }

  render() {
    const { value } = this.state;
    const { buttonProps, className, iconProps, variant } = this.props;

    let icon = '';

    switch (variant) {
      case 'check':
      default:
        icon = value ? ['fas', 'check-square'] : ['far', 'square'];
        break;
    }
    const classes = classNames(className, 'btn-toggle');

    return (
      <Button
        {...buttonProps}
        onClick={this.handleClick}
        className={classes}
        title="Click to toggle"
      >
        <FontAwesomeIcon size="lg" {...iconProps} icon={icon} />
      </Button>
    );
  }
}
