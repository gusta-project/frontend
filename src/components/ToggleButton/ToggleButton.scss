.btn.btn-toggle {
  padding: 0 8px;
}
