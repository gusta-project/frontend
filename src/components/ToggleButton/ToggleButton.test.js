import React from 'react';
import renderer from 'react-test-renderer';

import ToggleButton from './ToggleButton';

describe('<ToggleButton />', () => {
  const event = { test: true };
  const props = {
    onClick: jest.fn(),
    variant: 'check'
  };

  it('can render "check" variant', () => {
    const tree = renderer.create(<ToggleButton {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('uses "check" variant as default', () => {
    const component = renderer.create(<ToggleButton />);
    const instance = component.getInstance();

    expect(instance.props.variant).toEqual('check');
  });

  it('calls onClick on click', () => {
    const component = renderer.create(<ToggleButton {...props} />);
    const instance = component.getInstance();

    expect(props.onClick).not.toHaveBeenCalled();
    instance.handleClick(event);
    expect(props.onClick).toHaveBeenCalledWith(event);
  });

  it('updates state on click', () => {
    const component = renderer.create(<ToggleButton />);
    const instance = component.getInstance();

    expect(instance.state.value).toBe(false);
    instance.handleClick(event);
    expect(instance.state.value).toBe(true);
  });

  it('defaults to false value', () => {
    const component = renderer.create(<ToggleButton {...props} />);
    const instance = component.getInstance();

    expect(instance.state.value).toBe(false);
  });

  it('accepts initialValue prop', () => {
    const component = renderer.create(
      <ToggleButton {...props} initialValue={true} />
    );
    const instance = component.getInstance();

    expect(instance.state.value).toBe(true);
  });
});
