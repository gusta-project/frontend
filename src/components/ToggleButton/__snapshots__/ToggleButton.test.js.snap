// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButton /> can render "check" variant 1`] = `
<button
  className="btn-toggle btn btn-primary"
  disabled={false}
  onClick={[Function]}
  title="Click to toggle"
  type="button"
>
  <i
    className="fa far-square"
  />
</button>
`;
